Trashbin memory

jvillafanez opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.

Please set the following labels:

- Set label "3 - To review" for review or "2 - Development" in case the PR still has open tasks
- Set label "backport-request" if backport is needed
- Assignment: assign to self
- Milestone: set the same as the ticket this PR fixes, or "development" by default
- Reviewers: pick at least one
-->

## Description
Reduce memory usage in the TrashExpire background job

## Related Issue
https://github.com/owncloud/enterprise/issues/2925

## Motivation and Context
There have been problems related to excesive memory usage, probably caused by an abnormally big trashbin.
Note that problems with excesive group caching has also been detected, but the fix isn't implemented here yet. In addition, the `Trashbin::getLocations` calls should also be changed.

## How Has This Been Tested?
Manually tested with 3 users and a bunch of files inside a couple of folders. The code has been slightly changed to get the peak memory usage in several places.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
<!-- In case of incomplete PR, please list the open tasks here -->
- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

jvillafanez committed:
Reduce memory usage by using a generator when we fetch the trashed files

jvillafanez committed:
Clean cached preview information after being used

jvillafanez committed:
Adjust missing php docs

jvillafanez committed:
Remove unused variable to store exceptions

jvillafanez committed:
Adjust code style and fix type hints

micbar commented:
@jvillafanez @pako81 Is this code still needed? I see in the original ticket that you need a new ticket?
What about this one?
If valid, it would need more test coverage.

pako81 commented:
@micbar it is still under investigation. We are trying to verify if 10.2.0, which should already come with some improvements in terms of memory usage, is already solving the issue. If not, it may be required to apply this code and check the outcome.
@jvillafanez correct me if I am wrong

jvillafanez commented:
It's on my radar. I got side-tracked with other more urgent things.
The code is still needed because it fixes a couple of problems regarding high memory usage, but it's under investigation if we need more changes or not.

jvillafanez committed:
Include unittests

jvillafanez commented:
I don't think I can add more tests here:
* `deleteFileIfExpired` in "Trashbin.php" relies on the `self::delete` function, which access to the DB and creates 2 new View objects. In addition, it does a lot of things. Not only the dependencies can't be mocked, but also we'll need access both to the DB and the FS. Unittests are impossible to make here without a heavy refactor.
* `getTrashFilesGenerator` in "Helper.php". It was copied from the `getTrashFiles` and adjusted to use a generator. Most of the copied code can't be unittested. There is a new View object that can't be mocked, and the `Trashbin::getLocations` calls depends on access to the DB, so we can't mock that call.
* `run` in the "ExpireTrash.php" depends a couple of static code. Those calls aren't isolated and depend on several components, so we can't add unittests there.

The rest of the changes should be already covered by previous tests.

jvillafanez committed:
Adjust code style

jvillafanez commented:
Not sure how to proceed here. There are several problems with the tests due to a confusing architecture around the Files/Cache.

Options:
* Remove the change around Files/Cache/Cache (no more generator). We'll keep the rest of the changes. This will increase the memory usage, but it should fix CI and it also keep the changes at minimum.
* Spread the `getFolderContentsGenerator` and `getFolderContentsByIdGenerator` to all the Cache subclasses. I need to investigate if this is possible or not, mainly because of how generators work and the expected behaviour of the cache wrappers.
* Try to fix the CacheWrapper and subclasses by NOT inheriting from the Cache class. This probably will break things and we'll have additional things to fix.
* Try to implement a proper decorator pattern. The problem is that we probably won't be able to move from the current code to there without breaking a lot of things.

jvillafanez commented:
I think I'll close this PR and create a new one fixing the memory leaks found.

Although using a generator in the trashbin was a good idea to reduce memory usage, it requires changes in the Cache implementation which either needs to be in the interface or we need to downcast to a private implementation.

For consistency, if we downcast to a private implementation (which is wrong because the expected implementation is in core, not in the trashbin app), we'll also need to handle the rest of the implementation somehow, so we'll still need to rely on the interface. Sound better to rely on the interface in the first place.

If we decide to create a new method in the interface, all the implementations needs to provide something similar. This is starting to get out of scope. The additional problem is that the wrappers also need to implement the same method. This method have to return a generator which wraps the returned generator.
This is either reshaping the architecture or adding new functionality, but not fixing a bug.

jvillafanez commented:
Simplified version of the expected solution is in https://github.com/owncloud/core/pull/35612 
It doesn't have the changes for the trashbin, but the other 2 problems are also fixed there.

jvillafanez commented:
@micbar we should probably close this PR in favour of https://github.com/owncloud/core/pull/35612 . As said above, I don't think it's possible to fix the tests without major changes.

micbar commented:
@jvillafanez ok for me , please proceed as proposed :-)

jvillafanez commented:
Closing in favour of https://github.com/owncloud/core/pull/35612

phil-davis commented:
Closed, so backport not needed
